lms/models/user.py

```diff
@@ -9,7 +9,8 @@ class User(CreatedUpdatedMixin, BASE):
 
     __tablename__ = "user"
     __table_args__ = (
-        sa.UniqueConstraint("application_instance_id", "user_id", "h_userid"),
+        sa.UniqueConstraint("user_id", "application_instance_id"),
+        sa.UniqueConstraint("h_userid", "application_instance_id"),
```

Comment:
## About the ordering

Erg, this is a shame. I think the way it works is you can use a partial index as long as you start from the start and go forwards. So with user id's first it would allow you to efficiently ask which application instances a user id was in: probably not useful as the names are defined by the LMS. 

With the application instance first you would be able to ask for all users for an application instance. I'm not sure we'd use that either, but it sounds more likely to be something we care about.

Can you manually specify the index name here so you can have the name we want without having to change the default? If not, I don't think it's a big deal, and probably not worth rewriting the whole thing for, but on the other hand it's an annoying limitation that's forcing us to do sub-optimal things.

## About the constraints logically

I think this is right for what it's trying to do, but I'm not sure if what it's trying to do actually makes sense or not.

If the LMS id and application instance id are unique (i.e. we only have the first constraint, not the second), there can only be one row with that in there. So we can't have a situation like this:

 * `app_id: same, user_id: same, h_userid: same`
 * `app_id: same, user_id: same, h_userid: DIFFERENT`
 
Which seems like the main thing we care about: there's a unique record for each user per application instance. What the second condition protects against is:

 * `app_id: same, user_id: same, h_userid: same`
 * `app_id: same, user_id: DIFFERENT, h_userid: same`

What I'm trying to work out is if this is any of the DB's business or not? With the current method of generating `h` user ids from the other ids, I don't actually think this can happen. And if we want to change our minds and let it happen should the DB object?

I feel like this is easily argued from both sides:

 * If the code can't let this happen there's no point in the DB constraining it
 * If the code shouldn't let this happen, there's no harm in the DB double checking it

So why not belt and braces it and have it double checked by the DB? We can always drop the constraint later if we decide we want to redirect multiple users to the same account for some reason.

Reply:
To cut a long story short I think we could fix our `naming_convention` to avoid collisions like this but I don't think it's worth the trouble: our current `naming_convention` makes collisions rare enough and we can just manoeuvre around individual collisions like the one in this PR by specifying the constraint names manually: https://github.com/hypothesis/lms/pull/3494/commits/d4243efa18788a9370c9ef9f36962059f92e4912

(So I've changed both constraints to list the `application_instance_id` first.)